examples/step-79/step-79.cc

```diff
@@ -0,0 +1,2756 @@
+         FE_DGQ<dim>(0) ^ 1,
+         (FESystem<dim>(FE_Q<dim>(1) ^ dim)) ^ 1,
+         FE_DGQ<dim>(0) ^ 5)
+    , dof_handler(triangulation)
+    , density_ratio(.5)
+    , density_penalty_exponent(3)
+    , filter_r(.25)
+    , penalty_multiplier(1)
+    , timer(std::cout, TimerOutput::summary, TimerOutput::wall_times)
+  {}
+
+
+  // The first step then is to create the triangulation that matches
+  // the problem description in the introduction -- a 6-by-1
+  // rectangle (or a 6-by-1-by-1 box in 3d) where a force will be
+  // applied in the top center. This triangulation is then uniformly
+  // refined a number of times.
+  //
+  // In contrast to nearly the entire rest of this program, this
+  // function specifically assumes that we are in 2d and will
+  // require changes if we wanted to move to 3d simulations. We
+  // ensure that nobody tries to accidentally run in 3d without such
+  // modifications through an assertion at the top of the function.
+  template <int dim>
+  void SANDTopOpt<dim>::create_triangulation()
+  {
+    Assert(dim == 2, ExcNotImplemented());
+    GridGenerator::subdivided_hyper_rectangle(triangulation,
+                                              {6, 1},
+                                              Point<dim>(0, 0),
+                                              Point<dim>(6, 1));
+
+    triangulation.refine_global(3);
+
+    // The second step is to apply boundary indicators to parts of
+    // the boundary. The following code assigns boundary
+    // indicators to the bottom, top, left, and right boundaries
+    // of the box, respectively. The center region of the top
+    // boundary is given a separate boundary indicator: This is
+    // where we will apply the down force.
+    for (const auto &cell : triangulation.active_cell_iterators())
+      {
+        for (const auto &face : cell->face_iterators())
+          {
+            if (face->at_boundary())
+              {
+                const auto center = face->center();
+                if (std::fabs(center(1) - 0) < 1e-12)
+                  {
+                    face->set_boundary_id(BoundaryIds::no_force);
+                  }
+                else if (std::fabs(center(1) - 1) < 1e-12)
+                  {
+                    if ((std::fabs(center(0) - 3) < .3))
+                      {
+                        face->set_boundary_id(BoundaryIds::down_force);
+                      }
+                    else
+                      {
+                        face->set_boundary_id(BoundaryIds::no_force);
+                      }
+                  }
+                else if (std::fabs(center(0) - 0) < 1e-12)
+                  {
+                    face->set_boundary_id(BoundaryIds::no_force);
+                  }
+                else if (std::fabs(center(0) - 6) < 1e-12)
+                  {
+                    face->set_boundary_id(BoundaryIds::no_force);
+                  }
+              }
```

Comment:
Why don't you use the coloring flag of `subdivided_hyper_rectangle()`?

Reply:
yes, agreed. That would be much simpler.

Reply:
While looking into the coloring flag, I realized I actually only use this information for those parts of the boundary that get assigned the down_force boundary id. Would it still be worth doing the coloring when I shorten this to only assign that part of the boundary?